pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,450 @@
+        }
+      }
+
+      // Sort segments based on startTimeMs and endTimeMs in ascending order
+      TreeSet<OfflineSegmentZKMetadata> preSelectedSegmentsForOfflineTable = allSegmentsForOfflineTable.stream()
+          .filter(m -> selectedSegmentsBasedOnLineage.contains(m.getSegmentName()) && !segmentsScheduled.contains(
+              m.getSegmentName()))
+          .collect(Collectors.toCollection(() -> new TreeSet<>((a, b) -> {
+            long aStartTime = a.getStartTimeMs();
+            long bStartTime = b.getStartTimeMs();
+            return aStartTime != bStartTime ? Long.compare(aStartTime, bStartTime)
+                : Long.compare(a.getEndTimeMs(), b.getEndTimeMs());
+          })));
+
+      if (preSelectedSegmentsForOfflineTable.isEmpty()) {
+        LOGGER.warn("Skip generating task: {} for table: {}, no segment is found to merge.", taskType,
+            offlineTableName);
+        continue;
+      }
+
+      // Sort merge levels based on bucket time period
+      Map<String, Map<String, String>> mergeLevelToConfigs = MergeRollupTaskUtils.getLevelToConfigMap(taskConfigs);
+      List<Map.Entry<String, Map<String, String>>> sortedMergeLevelConfigs = mergeLevelToConfigs.entrySet()
+          .stream()
+          .sorted(Comparator.comparingLong(
+              e -> TimeUtils.convertPeriodToMillis(e.getValue().get(MinionConstants.MergeTask.BUCKET_TIME_PERIOD_KEY))))
+          .collect(Collectors.toList());
+
+      // Schedule tasks from lowest to highest merge level
+      String lowerMergeLevel = null;
+      long lowerMergeLevelWatermarkMs = -1;
+
+      for (Map.Entry<String, Map<String, String>> entry : sortedMergeLevelConfigs) {
+        String mergeLevel = entry.getKey();
+        Map<String, String> mergeConfig = entry.getValue();
+
+        // Get the bucket size and buffer
+        long bucketMs =
+            TimeUtils.convertPeriodToMillis(mergeConfig.get(MinionConstants.MergeTask.BUCKET_TIME_PERIOD_KEY));
+        long bufferMs =
+            TimeUtils.convertPeriodToMillis(mergeConfig.get(MinionConstants.MergeTask.BUFFER_TIME_PERIOD_KEY));
+
+        // Get watermark from MergeRollupTaskMetadata ZNode.
+        // windowStartMs = watermarkMs, windowEndMs = windowStartMs + bucketTimeMs
+        long waterMarkMs =
+            getWatermarkMs(offlineTableName, preSelectedSegmentsForOfflineTable.first().getStartTimeMs(), bucketMs,
+                mergeLevel);
+        long windowStartMs = waterMarkMs;
+        long windowEndMs = windowStartMs + bucketMs;
+
+        // Skip scheduling if there's incomplete task for current mergeLevel
+        if (inCompleteMergeLevelToTaskConfig.containsKey(mergeLevel)) {
+          LOGGER.info(
+              "Found incomplete tasks: {} of merge level: {} for the same table: {}, Skipping task generation: {}",
+              inCompleteMergeLevelToTaskConfig.get(mergeLevel), mergeLevel, offlineTableName, taskType);
+          lowerMergeLevel = mergeLevel;
```

Comment:
The logic is a little bit mixed here. I would recommend do multiple passes. First pass we just calculate the watermark for each merge level, and persist them back. Second path we calculate the segments to merge for each level

Reply:
As discussed offline, the segment selection is relatively straight forward compared with window/watermark calculation, it should be okay to maintain them in one pass. Updated the comments to list all terminate conditions of the for loop, hope it helps to make the logic clearer.